commit 753461e4dbcfa3e6b224526108dd2409571652ee (refs/changes/82/1782/3)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-04-13T01:13:04+00:00 (2 years, 6 months ago)
    
    joyent/node-cueball#107 Bootstrap resolver not stopped when the last resolver using it stops
    Reviewed by: David Pacheco <dap@joyent.com>

diff --git a/CHANGES.adoc b/CHANGES.adoc
index 758db06..9eb3d07 100644
--- a/CHANGES.adoc
+++ b/CHANGES.adoc
@@ -6,6 +6,15 @@ toc::[]
 
 ## v2.x
 
+### v2.2.5
+
+Maintenance release.
+
+Bugs fixed:
+
+ - #107 Bootstrap resolver not stopped when the last resolver using it
+   stops
+
 ### v2.2.4
 
 Maintenance release.
diff --git a/lib/resolver.js b/lib/resolver.js
index a7f456d..c19e59a 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -262,6 +262,7 @@ function CueBallDNSResolver(options) {
 		this.r_service = '_dns._udp';
 		this.r_defport = 53;
 		this.r_maxres = 10;
+		this.r_refCount = 0;
 		this.setMaxListeners(500);
 	}
 
@@ -364,6 +365,11 @@ CueBallDNSResolver.prototype.list = function () {
 
 CueBallDNSResolver.prototype.state_init = function (S) {
 	this.r_stopping = false;
+	if (this.r_bootstrap !== undefined) {
+		if (--this.r_bootstrap.r_refCount <= 0)
+			this.r_bootstrap.stop();
+		this.r_bootstrap = undefined;
+	}
 	S.on(this, 'startAsserted', function () {
 		S.gotoState('check_ns');
 	});
@@ -394,6 +400,7 @@ CueBallDNSResolver.prototype.state_check_ns = function (S) {
 			CueBallDNSResolver.bootstrapResolvers[notIp[0]] =
 			    this.r_bootstrap;
 		}
+		++this.r_bootstrap.r_refCount;
 		S.gotoState('bootstrap_ns');
 	} else {
 		mod_fs.readFile('/etc/resolv.conf', 'ascii',
diff --git a/package.json b/package.json
index 1adfa25..067bb6e 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "cueball",
-  "version": "2.2.4",
+  "version": "2.2.5",
   "description": "manage a pool of connections to a multi-node service where nodes are listed in DNS",
   "main": "lib/index.js",
   "dependencies": {
